### PR TITLE
[2.1] Create proxies for base type and derived types in any order

### LINKED ISSUE #11886: Lazy Loading on Inherited type not working
Lazyloading - when using a base type which is inherited by other types is not working

```
Exception message: "The property or navigation 'LazyLoader' cannot be added to the entity type 'Entity' because a property or navigation with the same name already exists on entity type 'Company'."
Stack trace: "   at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityType.AddServiceProperty(MemberInfo memberInfo, ConfigurationSource configurationSource)\r\n   at Microsoft.EntityFrameworkCore.Proxies.Internal.ProxyBindingRewriter.Apply(InternalModelBuilder modelBuilder)\r\n   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelBuilt(InternalModelBuilder modelBuilder)\r\n   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)\r\n   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)\r\n   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)\r\n   at System.Lazy`1.CreateValue()\r\n   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()\r\n   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()\r\n   at Microsoft.EntityFrameworkCore.Internal.InternalAccessorExtensions.GetService[TService](IInfrastructure`1 accessor)\r\n   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_DatabaseCreator()\r\n   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()\r\n   at ConsoleApp1.Program.<MainAsync>d__1.MoveNext() in C:\\Sources\\Playground\\NetCore.M2MTest\\NetCore.M2MTest\\Program.cs:line 23\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at ConsoleApp1.Program.Main(String[] args) in C:\\Sources\\Playground\\NetCore.M2MTest\\NetCore.M2MTest\\Program.cs:line 15"
```

### Steps to reproduce

```c#
    public class Entity
    {
        public int ID { get; set; }
    }

    public class Company : Entity
    {
        public string CompanyTitle { get; set; }
    }

    public class Person : Entity
    {
        public string FirstName { get; set; }
    }

    modelBuilder.Entity<Entity>()
        .ToTable("Entities")
        .HasDiscriminator<string>("Discriminator")
        .HasValue<Company>("Company")
        .HasValue<Person>("Person");

    options.UseLazyLoadingProxies().UseSqlServer("Server=(LocalDb)\\MSSQLLocalDB;Database=TestDB;Trusted_Connection=true;MultipleActiveResultSets=true");

```

### Further technical details
EF Core version: 2.1.0-preview2-final
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.7 (Preview)

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #11886

### CODE REVIEW

File: src/EFCore.Proxies/Proxies/Internal/ProxyBindingRewriter.cs [hunk 1]
```diff
@@ -63,6 +63,13 @@ public virtual InternalModelBuilder Apply(InternalModelBuilder modelBuilder)
 
                         var proxyType = _proxyFactory.CreateLazyLoadingProxyType(entityType);
 
+                        foreach (var conflictingProperty in entityType.GetDerivedTypes()
```
Comment (smitpatel) [inline src/EFCore.Proxies/Proxies/Internal/ProxyBindingRewriter.cs]: Do we want to follow Detach/Attach pattern which `InternalEntityTypeBuilder` does for adding a property in base type with conflicting derived type property?

Comment (ajcvickers) [reply src/EFCore.Proxies/Proxies/Internal/ProxyBindingRewriter.cs]: Not sure what this means?

Comment (smitpatel) [reply src/EFCore.Proxies/Proxies/Internal/ProxyBindingRewriter.cs]: https://github.com/aspnet/EntityFrameworkCore/blob/14290a0f71f38c727deaa76f8bca326321761610/src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs#L389-L394

Detach derived properties, create new base property, re-attach derived property to merge config.

Comment (AndriySvyryd) [reply src/EFCore.Proxies/Proxies/Internal/ProxyBindingRewriter.cs]: I agree with @ajcvickers, reattaching is only needed if there was some user configuration we wanted to preserve.

**Comment by ajcvickers** [comment]
@bricelam @natemcmaster Linux build is failing to find the dotnet-ef tool. Is this a known issue?
```
13:48:49 /mnt/j/workspace/aspnet_EntityFrameworkCore/release_2.1/linux-Configuration_Release_prtest/src/dotnet-ef/dotnet-ef.csproj : error : Unable to resolve 'Microsoft.DotNet.GlobalTools.Sdk (>= 2.1.0-rtm-15782)' for '.NETStandard,Version=v0.0'.
13:48:49 /mnt/j/workspace/aspnet_EntityFrameworkCore/release_2.1/linux-Configuration_Release_prtest/src/dotnet-ef/dotnet-ef.csproj : error MSB4236: The SDK 'Microsoft.DotNet.GlobalTools.Sdk' specified could not be found.
13:48:50 
13:48:50 Build FAILED.
13:48:50 
13:48:50 /mnt/j/workspace/aspnet_EntityFrameworkCore/release_2.1/linux-Configuration_Release_prtest/src/dotnet-ef/dotnet-ef.csproj : error : Unable to resolve 'Microsoft.DotNet.GlobalTools.Sdk (>= 2.1.0-rtm-15782)' for '.NETStandard,Version=v0.0'.
13:48:50 /mnt/j/workspace/aspnet_EntityFrameworkCore/release_2.1/linux-Configuration_Release_prtest/src/dotnet-ef/dotnet-ef.csproj : error MSB4236: The SDK 'Microsoft.DotNet.GlobalTools.Sdk' specified could not be found.
13:48:50     0 Warning(s)
13:48:50     2 Error(s)
```

**Comment by natemcmaster** [comment]
No, not known. I haven't seen this before and couldn't figure out why, especially when all other CI configs appear to be working. If it persists, you might try updating the korebuild version. Run `build.cmd -u /t:Cow` and commit the korebuild-lock.txt file

@dotnet-bot test Ubuntu 16.04 Release Build please

**Pull request merged**